### PR TITLE
[WIP] Adds support for heroku formation management

### PR DESCRIPTION
**Pull request by dalehamel**
This adds support for managing heroku formations (dynos and sizes). It might make more sense to call this 'heroku dynos', but formations the more correct term.

Notably, this somewhat violates the CRUD promise, as there is actually no way to actually create dynos for apps that have not yet been deployed. Once an app is deployed however, this will allow for its dyno formations to be managed in terraform. This need to deploy an app makes this tricky to test, however, the code is really quite simple - just a few calls and comparisons.

The workflow I hope to support here is to manage heroku dynos via git. This would allow us to code review dyno changes, and have a history of who changed them and why. The use case here is that rather than giving every developer access to every heroku account, and trying to somehow track who changed what, we just have a single account that can manage all dynos - elegantly tracked in Git.

Since dynos are a pretty key part of heroku, and I cannot see any other way to manage them with terraform, I feel there is a fundamental need for this support - that's why I've made this patch.

In my tests, I can confirm:

* The app is successfully created without the any formations specified
* Future plans / applies will continue to show that the formations are outstanding
* Once the app is deployed, and dynos exist for the given formation, the formation is successfully created.
* The formation can successfully manage the size and quantity of dynos for a given type of the app

### CODE REVIEW

**Comment by dalehamel** [comment]
@mitchellh @pearkes @catsby for review since your names come up as heroku provider contribs :)

**Comment by catsby** [comment]
Thanks @dalehamel ! This is the first I've heard of "formations", could you please link me some relevant Devcenter  / API docs?

**Comment by dalehamel** [comment]
Mostly from https://devcenter.heroku.com/articles/platform-api-reference#formation for the api, and the heroku library you are already using supports this

**Comment by dalehamel** [comment]
@catsby other general docs on formations : https://devcenter.heroku.com/articles/scaling#dyno-formation, specifically:

> The term dyno formation refers to the layout of your app’s dynos at a given time. The default formation for simple apps will be a single web dyno, whereas more demanding applications may consiste of web, worker, clock etc… process types. In the examples above, the formation was first changed to two web dynos, then two web dynos and a worker.

tl;dr you define process types in your Procfile (sadly, this cannot be done through the API or we'd have an end-to-end solution). Each process type gets its own 'formation' of dynos - the quantity and size of the dyno.

This allows you to manage those formations. Workflow looks like:

1. Terraform creates your app (you may specify formations for your process types now, but it won't do anything)
2. You ship your app to heroku with a procfile containing your process types
3. Terraform manages the formations of dynos for those process types on every successive apply / plan.

**Comment by dalehamel** [comment]
If this ships, i'd be happy to write docs for it - where is that done? Is it github.io?

**Comment by dalehamel** [comment]
@catsby anything I can do to assist your review?

**Comment by dalehamel** [comment]
bump @catsby @mitchellh and @phinze because you took a look at my other PR :)

**Comment by dalehamel** [comment]
sorry to bump again @catsby is anyone looking at this? It's the last pending patch we have against terraform, we'd prefer not to run a forked version, but we need support for managing heroku dyno formations.

**Comment by catsby** [comment]
Hey @dalehamel I apologize for the delay here, I've been dug into some other problems. 

We'll need documentation for this yes. They are simple markdown files in the `website` directory, ex:

- https://github.com/hashicorp/terraform/blob/master/website/source/docs/providers/heroku/r/app.html.markdown

you can either `cd` into that directory and use `bundle install`, then `bundle exec middleman` to get a local preview, or I believe there's a working Vagrant file to assist in there as well. If nothing else, you can just push the markdown and preview it in this pull request. 

We'll need those first before we can merge, but I'll review the code and maybe take it for a spin now

**Comment by catsby** [comment]
This looks a bit out of date, can you try merging/rebasing with `master`? Also, please run `go fmt` on the files.

File: builtin/providers/heroku/resource_heroku_formation.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+func resourceHerokuFormationDelete(d *schema.ResourceData, meta interface{}) error {
+	d.SetId("")
+	return nil
+}
+
+func resourceHerokuFormationExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	client := meta.(*heroku.Service)
+
+	formations, err := client.FormationList(d.Get("app").(string), (&heroku.ListRange{Max: 10000}))
```
Comment (catsby) [inline builtin/providers/heroku/resource_heroku_formation.go]: `(&heroku.ListRange{Max: 10000})`

The parenthesis are not needed here. Is this `10000` a practical limit to specify? It looks like the maximum is `1,000`, unless I'm reading [Ranges][1] wrong

[1]: https://devcenter.heroku.com/articles/platform-api-reference#ranges

File: builtin/providers/heroku/resource_heroku_formation.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+func resourceHerokuFormationUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*heroku.Service)
+
+	app := d.Get("app").(string)
+
+	quantity := d.Get("quantity").(int)
+	size := d.Get("size").(string)
+
+	if d.Id() == d.Get("type") && (d.HasChange("quantity") || d.HasChange("size")) {
```
Comment (catsby) [inline builtin/providers/heroku/resource_heroku_formation.go]: If we call `d.SetId("")` in the `create` method, when will `d.Id() == d.Get("type")` ever evaluate to true?

Comment (catsby) [inline builtin/providers/heroku/resource_heroku_formation.go]: ah, I guess that gets set later in the `read` method.

Comment (catsby) [inline builtin/providers/heroku/resource_heroku_formation.go]: following that loop though, if `d.SetId()` gets called with the formation id from Heroku in the `read` method, when then will `d.Id() == d.Get("type")` evaluate to `true`?

**Comment by catsby** [comment]
Can you give me some example config or steps to test this out? If I have an app already running, for example

**Comment by dalehamel** [comment]
Thanks for the review!

I'll get you a sample ASAP, I'm just hopping between air ports today.

More to follow.

File: builtin/providers/heroku/resource_heroku_formation.go (hunk 2 above)
Comment (dalehamel) [inline builtin/providers/heroku/resource_heroku_formation.go]: it evaluates to true when there is a formation of the type specified on heroku.

**Comment by dalehamel** [comment]
@catsby some sample code:

```
variable "type" { default = "web" } // from procfile
variable "quantity" { default = 0 }
variable "tier" { default = "0" }

// Available tiers - do not call directly, hack to get a hash
variable "tiers" {
  default = {
    "0" = "1X"
    "1" = "2X"
    "2" = "PX"
  }
}

resource "heroku_formation" "default" {
  app  = "${var.app}"
  type = "${var.type}"
  quantity = "${var.quantity}"
  size = "${lookup(var.tiers, var.tier)}"
}
```

This will manage the dynos in the 'web' formation. To use this:

* Create a heroku app with terraform
* Ship some code to that heroku app, with a procfile specifying the dyno formations you want to manage (web, job, etc)
* Specify the dyno formations you want to manage using the above code:
 * For each formation (web,job etc) you can now control the size and quantity of dynos.

**Comment by dalehamel** [comment]
bump @catsby @phinze

**Comment by catsby** [comment]
Hey @dalehamel – I checked out your branch and tried this out but I can't get it to work correctly. 

I have a Heroku app scaled to 0 dynos, and I'm trying to apply the config you've shared (except I specify `quantity = 1`). Terraform shows a diff and tries to apply, but nothing is scaled up, and Terraform still shows a diff.
 
Here's a gist of what I tried: https://gist.github.com/catsby/9bd1b9472e98a750f88d

Reading the code, it seems that the first pass doesn't do any creation or setup from the config. It passes from `create` to `update`, where `update` does nothing because `d.Id()` isn't set yet, and then on to `read` where we read existing information(?). 

Investigating further, this block never evaluates to `true`:

    if d.Id() == d.Get("type") && (d.HasChange("quantity") || d.HasChange("size")) {

Because you're setting `d.Id()` above to be the formation id.

**Comment by dalehamel** [comment]
@catsby there's no way to create a 'proc' in the heroku api. You must deploy an app to heroku with a procfile. Once that is done, that proc can be managed with a formation.

**Comment by catsby** [comment]
I created a Heroku application, including a `Procfile`. I don't see the relevance of that though, what am I missing? You can see my sample app here, with TF file and state included: 

- https://github.com/catsby/golang-heroku

Notice in the `statefile` the `id` is `1e03f01f-430d-4035-a639-e816a860a857`, and not `web`, so that update block is never entered

**Comment by catsby** [comment]
Hey @dalehamel – any idea on my last question there? I took this for a spin but it didn't quite work the way I was expecting.

**Comment by fgrehm** [comment]
Hey @dalehamel do u need any help wrapping this up? I _might_ be able to jump in to help :smile: :beers:

**Comment by dalehamel** [comment]
@fgrehm i think it just needs a rebase and another pass.

One of the fundamentally limitting things here is that heroku will not let you manage dynos until the app has been deployed. It's kind of weird, i think you should be able to pay for resources you aren't using (yet), but I guess they differ on that.

For that reason, you cannot fully converge the formation in one step.

You can do this:

1. Build the app, optionally specify the formations. If you do, they won't take effect yet.
2. Deploy the app
3. specify the formation, or just re-run terraform apply.

@catsby is that what you tried?

**Comment by catsby** [comment]
Yes. I still saw no affect, the gist I shared outlines what I saw. I have some questions above that I need more information on.

**Comment by phinze** [comment]
Hey folks! Just checking in. @dalehamel @fgrehm are either of you able to pick up the last bit of work to land this?

**Comment by fgrehm** [comment]
Sorry but not in the short term. I'm in between jobs and I'm not sure if / when I'll have a need for this :cry:

**Comment by stack72** [comment]
Hi all

there hasn't been any activity on this PR for > 60 days now. I am going to close it for now. Please feel free to reopen if you want to start working on it again. Thank you so much for the contribution made on this

Paul
